lib/Maskings/Maskings.cpp

```diff
@@ -0,0 +1,309 @@
+
+    if (res.status != ParseResult<Maskings>::VALID) {
+      return MaskingsResult(MaskingsResult::ILLEGAL_DEFINITION, res.message);
+    }
+
+    return MaskingsResult(std::move(maskings));
+  } catch (velocypack::Exception const& e) {
+    std::string msg =
+        "cannot parse maskings file '" + filename + "': " + e.what();
+    LOG_TOPIC(DEBUG, Logger::CONFIG) << msg << ". file content: " << definition;
+
+    return MaskingsResult(MaskingsResult::CANNOT_PARSE_FILE, msg);
+  }
+}
+
+ParseResult<Maskings> Maskings::parse(VPackSlice const& def) {
+  if (!def.isObject()) {
+    return ParseResult<Maskings>(ParseResult<Maskings>::DUPLICATE_COLLECTION,
+                                 "expecting an object for masking definition");
+  }
+
+  for (auto const& entry : VPackObjectIterator(def, false)) {
+    std::string key = entry.key.copyString();
+    LOG_TOPIC(TRACE, Logger::CONFIG) << "masking collection '" << key << "'";
+
+    if (_collections.find(key) != _collections.end()) {
+      return ParseResult<Maskings>(ParseResult<Maskings>::DUPLICATE_COLLECTION,
+                                   "duplicate collection entry '" + key + "'");
+    }
+
+    ParseResult<Collection> c = Collection::parse(entry.value);
+
+    if (c.status != ParseResult<Collection>::VALID) {
+      return ParseResult<Maskings>(
+          (ParseResult<Maskings>::StatusCode)(int)c.status, c.message);
+    }
+
+    _collections[key] = c.result;
+  }
+
+  return ParseResult<Maskings>(ParseResult<Maskings>::VALID);
+}
+
+bool Maskings::shouldDumpStructure(std::string const& name) {
+  auto const itr = _collections.find(name);
+
+  if (itr == _collections.end()) {
+    return false;
+  }
+
+  switch (itr->second.selection()) {
+    case CollectionSelection::FULL:
+      return true;
+    case CollectionSelection::IGNORE:
+      return false;
+    case CollectionSelection::STRUCTURE:
+      return true;
+  }
+}
+
+bool Maskings::shouldDumpData(std::string const& name) {
+  auto const itr = _collections.find(name);
+
+  if (itr == _collections.end()) {
+    return false;
+  }
+
+  switch (itr->second.selection()) {
+    case CollectionSelection::FULL:
+      return true;
+    case CollectionSelection::IGNORE:
+      return false;
+    case CollectionSelection::STRUCTURE:
+      return false;
+  }
+}
+
+VPackValue Maskings::maskedItem(Collection& collection,
+                                std::vector<std::string>& path,
+                                std::string& buffer, VPackSlice const& data) {
+  static std::string xxxx("xxxx");
+
+  if (path.size() == 1) {
+    if (path[0] == "_key" || path[0] == "_id" || path[0] == "_rev") {
+      if (data.isString()) {
+        velocypack::ValueLength length;
+        char const* c = data.getString(length);
+        buffer = std::string(c, length);
+        return VPackValue(buffer);
+      } else if (data.isInteger()) {
+        return VPackValue(data.getInt());
+      }
```

Comment:
Should we also handle UInt and Double values here?

Reply:
In ArangoDB all these values will be strings... Not sure if it is necessary to check for non-string values here.